ion-schema/src/result/invalid_schema_error.rs

```diff
@@ -0,0 +1,165 @@
+use crate::result::{HasIslSourceLocation, IonSchemaError, IonSchemaResult, IslSourceLocation};
+use std::collections::BTreeSet;
+use std::fmt::{Display, Formatter, Write};
+use thiserror::Error;
+
+/// Collects multiple errors to report them together.
+///
+/// ```ignore
+/// # use ion_schema::IonSchemaResult;
+/// #
+/// use ion_schema::result::ErrorCollector;
+///
+/// fn error_prone_function(i: i8) -> IonSchemaResult<()> {
+///     todo!()
+/// }
+/// fn some_fun() -> IonSchemaResult<()> {
+///     // Create the ErrorCollector
+///     let mut errors = ErrorCollector::default();
+///     for i in 1..10 {
+///         let Some(ok_value) = errors.ok_or_push_err(error_prone_function(i)) else {
+///             continue;
+///         }
+///         // Do something with `ok_value`
+///     }
+///     // Transform the error collector into a result.
+///     errors.into_result_with(())
+/// }
+/// ```
+#[derive(Debug, Default)]
+pub(crate) struct InvalidSchemaErrorCollector {
+    errors: BTreeSet<InvalidSchemaDetails>,
```

Comment:
(Question) Why are we using a `BTreeSet` here. I don't think the errors collected has to be sorted unless I am missing something here.

Reply:
There are two different parts to this question:

> I don't think the errors collected has to be sorted

In order to have a better user experience, they are sorted by `schema_id`, `row`, and `column`. Even though reading the ISL may occur in order and errors would already be sorted, resolving the imports and type references is a separate step (and itself includes multiple steps) so errors are not sorted by default.

> Why are we using a BTreeSet here

Because I didn't think it was appropriate to allocate a vec to hold a sorted copy of the error details in the `Display` impl. The other option is to sort them as they are aggregated, for which the `BTreeSet` is the natural choice. This, however, is an implementation detail that can change if necessary.